File: compiler/src/Span.swift
public struct Span: Sendable, CustomStringConvertible, Comparable {
    public let path: String
    public let start: Location
    public let end: Location
    public let source: Substring

    public init(path: String, start: Location, end: Location, in source: String) {
        let startIndex = source.utf16.index(source.startIndex, offsetBy: start.index)
        let endIndex = source.utf16.index(source.startIndex, offsetBy: end.index)

        self.path = path
        self.start = start
        self.end = end
        self.source = Substring(source.utf16[startIndex..<max(startIndex, endIndex)])
    }

    public static var empty: Span { Span(path: "", start: .empty, end: .empty, in: "") }

    public func joined(with other: Self, in source: String) -> Self {
        .init(path: self.path, start: self.start, end: other.end, in: source)
    }

    public func isInside(_ other: Self) -> Bool {
        self.path == other.path && self.start.index >= other.start.index
            && self.end.index <= other.end.index
    }

    public var description: String {
        "\(self.path):\(self.start.line):\(self.start.column)-\(self.end.line):\(self.end.column)"
    }

    public static func < (lhs: Self, rhs: Self) -> Bool {
        if lhs.path != rhs.path { return lhs.path < rhs.path }
        if lhs.start != rhs.start { return lhs.start < rhs.start }
        return lhs.end < rhs.end
    }
}

public struct Location: Sendable, Encodable, Comparable {
    public let line: Int
    public let column: Int
    public let index: Int

    public static var empty: Location { Location(line: 1, column: 1, index: 0) }

    public static func < (lhs: Self, rhs: Self) -> Bool { lhs.index < rhs.index }
}

File: compiler/core/src/span.rs
use serde::{Deserialize, Serialize};
use smol_str::SmolStr;
use std::fmt::Display;

pub type Str = SmolStr;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub struct Span {
    pub path: Str,
    pub start: Location,
    pub end: Location,
    pub source: Str,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub struct Location {
    pub line: usize,
    pub column: usize,
    pub index: usize,
}

impl Span {
    pub fn new_in(path: Str, start: Location, end: Location, source: Str) -> Self {
        let source = Str::new(&source[start.index..start.index.max(end.index)]);

        Span {
            path,
            start,
            end,
            source,
        }
    }

    pub fn empty() -> Self {
        Span::new_in(
            Str::new_static(""),
            Location::empty(),
            Location::empty(),
            Str::new_static(""),
        )
    }

    pub fn join_in(&self, other: &Self, source: Str) -> Self {
        Span::new_in(
            self.path.clone(),
            self.start.clone(),
            other.end.clone(),
            source,
        )
    }
}

impl Display for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}:{}:{}-{}:{}",
            self.path, self.start.line, self.start.column, self.end.line, self.end.column
        )
    }
}

impl Location {
    pub fn empty() -> Self {
        Location {
            line: 1,
            column: 1,
            index: 0,
        }
    }
}
